version: '3.8' 

services:   
  caddy:    
    container_name: swapi_caddy   
    image: caddy:2.6-alpine    
    ports:      
      - "80:80"      
      - "443:443"    
    volumes:      
      - ./docker/caddy/Caddyfile:/etc/caddy/Caddyfile      
      - caddy_data:/data      
      - caddy_config:/config    
    networks:      
      - swapi_network    
    depends_on:      
      - app    
    restart: always   
    
  mysql:    
    container_name: swapi_mysql   
    image: mysql:latest    
    environment:      
      - MYSQL_ROOT_PASSWORD=${MYSQL_ROOT_PASSWORD}      
      - MYSQL_DATABASE=${DB_NAME}      
      - MYSQL_USER=${DB_USER}      
      - MYSQL_PASSWORD=${DB_PASS}    
    expose:      
      - "3306"    
    volumes:      
      - ./docker/mysql/data:/var/lib/mysql    
    networks:      
      - swapi_network    
    healthcheck:      
      test: ["CMD-SHELL", "mysqladmin ping -h localhost -p${MYSQL_ROOT_PASSWORD} || exit 1"]      
      interval: 10s      
      timeout: 5s      
      retries: 5    
    restart: on-failure   
   
  app:    
    container_name: swapi_app   
    build:      
      context: .      
      dockerfile: Dockerfile.prod      
      args:        
        - NODE_ENV=production        
        - ENV_FILE=.env.production    
    environment:      
      - NODE_ENV=production      
      - DB_HOST=mysql      
      - DB_PORT=${DB_PORT}      
      - APP_PORT=${APP_PORT}      
      - DB_NAME=${DB_NAME}      
      - DB_USER=${DB_USER}      
      - DB_PASS=${DB_PASS}    
    working_dir: /usr/src/app    
    volumes:      
      - ./:/usr/src/app    
    expose:      
      - "3000"    
    networks:      
      - swapi_network    
    depends_on:      
      mysql:        
        condition: service_healthy    
    restart: always 
    
networks: 
  swapi_network:
    driver: bridge 

volumes:   
  caddy_data:  
  caddy_config: